### PR TITLE
exp/services/ledgerexporter: create go integration tests for sub commands

### PR DESCRIPTION
**Pull request by sreuland**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [x] This PR adds tests for the most critical parts of the new functionality or fixes.
* [x] I've updated any docs ([developer docs](https://developers.stellar.org/api/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What

Added new go integration tests that verify ledger exporter 'scan-and-fill' and 'append' commands. 
- uses quickstart for standalone core
- uses fsouza/fake-gcs-server
- uses `support/datastore` to verify exported files exist on gcs.

```
$ LEDGEREXPORTER_INTEGRATION_TESTS_ENABLED=true \
LEDGEREXPORTER_INTEGRATION_TESTS_CAPTIVE_CORE_BIN=stellar-core \
go test -v -race -run TestLedgerExporterTestSuite ./exp/services/ledgerexporter/...
```

### Why

refactor from docker based testing infrastructure, follow standard convention in mono-repo of integration tests being next to unit tests, but enabled with env variable.

Closes https://stellarorg.atlassian.net/browse/HUBBLE-469

### Known limitations

not fully parallelizable yet, some measure were taken for dynamic port usage of fake gcs server.

### CODE REVIEW

File: exp/services/ledgerexporter/README.md [hunk 1]
```diff
@@ -128,3 +128,18 @@ Arguments:
 - `--end <end_ledger>` (required): The ending ledger sequence number in the range.
 - `--config-file <config_file_path>` (optional): The path to your configuration file, containing details like GCS bucket information. If not provided, the exporter will look for config.toml in the directory where you run the command.
 
+#### Running Integration Tests:
```
Comment (urvisavla) [inline exp/services/ledgerexporter/README.md:131]: This should be moved to the DEVELOPER_GUIDE.md

File: exp/services/ledgerexporter/README.md [hunk 2]
```diff
@@ -128,3 +128,18 @@ Arguments:
...
+
+`LEDGEREXPORTER_INTEGRATION_TESTS_ENABLED=true` is required environment variable to allow
+tests to run.
+
+Optional, tests will try to run `stellar-core` from o/s PATH for captive core, if not resolvable, then set `LEDGEREXPORTER_INTEGRATION_TESTS_CAPTIVE_CORE_BIN=/path/to/stellar-core` 
+
+Optional, can override the version of quickstart used to run standalone stellar network, p `LEDGEREXPORTER_INTEGRATION_TESTS_QUICKSTART_IMAGE=docker.io/stellar/quickstart:<tag>`. Ideally don't need to change this, but is available.
+
+```
+$ LEDGEREXPORTER_INTEGRATION_TESTS_ENABLED=true go test -v -race -run TestLedgerExporterTestSuite ./exp/services/ledgerexporter/...
```
Comment (urvisavla) [inline exp/services/ledgerexporter/README.md]: Getting the following error when running integration tests: 
"could not create quickstart docker container, stellar/quickstart, error: Error response from daemon: No such image: stellar/quickstart" 

Do I need to explicitly pull the quickstart Docker image? If so, we should include this step in the doc.

File: exp/services/ledgerexporter/README.md [hunk 3]
```diff
@@ -128,3 +128,18 @@ Arguments:
...
+#### Running Integration Tests:
+from top directory of stellar/go repo, run go test to launch ledger exporter integration
+tests.
+
+`LEDGEREXPORTER_INTEGRATION_TESTS_ENABLED=true` is required environment variable to allow
+tests to run.
+
+Optional, tests will try to run `stellar-core` from o/s PATH for captive core, if not resolvable, then set `LEDGEREXPORTER_INTEGRATION_TESTS_CAPTIVE_CORE_BIN=/path/to/stellar-core` 
+
+Optional, can override the version of quickstart used to run standalone stellar network, p `LEDGEREXPORTER_INTEGRATION_TESTS_QUICKSTART_IMAGE=docker.io/stellar/quickstart:<tag>`. Ideally don't need to change this, but is available.
```
Comment (urvisavla) [inline exp/services/ledgerexporter/README.md]:
```suggestion
Optional, can override the version of quickstart used to run standalone stellar network,  `LEDGEREXPORTER_INTEGRATION_TESTS_QUICKSTART_IMAGE=docker.io/stellar/quickstart:<tag>`. Ideally don't need to change this, but is available.
```

File: exp/services/ledgerexporter/DEVELOPER_GUIDE.md [hunk 4]
```diff
@@ -32,14 +32,27 @@ To achieve its goals, the ledger exporter uses the following architecture, which
...
 
-## Build, Run and Test using Docker
+## Build and Run using Docker
 The Dockerfile contains all the necessary dependencies (e.g., Stellar-core) required to run the ledger exporter. 
 
 - Build: To build the Docker container, use the provided [Makefile](./Makefile). Simply run make `make docker-build` to build a new container after making any changes.
 
 - Run: For instructions on running the Docker container, refer to the [Installation Guide](./README.md).
 
-- Test: To test the Docker container, refer to the [docker-test](./Makefile) command for an example of how to use the [GCS emulator](https://github.com/fsouza/fake-gcs-server) for local testing.
```
Comment (urvisavla) [inline exp/services/ledgerexporter/DEVELOPER_GUIDE.md:42]: Thanks for updating.

File: .github/workflows/ledgerexporter-release.yml [hunk 5]
```diff
@@ -10,17 +10,37 @@ jobs:
     name: Test and push the Ledger Exporter images
     runs-on: ubuntu-latest
     env:
+      LEDGEREXPORTER_INTEGRATION_TESTS_ENABLED: "true"
+      LEDGEREXPORTER_INTEGRATION_TESTS_CAPTIVE_CORE_BIN: /usr/bin/stellar-core
+      LEDGEREXPORTER_INTEGRATION_TESTS_QUICKSTART_IMAGE: docker.io/stellar/quickstart:testing
       STELLAR_CORE_VERSION: 21.1.0-1921.b3aeb14cc.focal
-      VERSION: ${GITHUB_REF_NAME#ledgerexporter-v}
+      VERSION: ${GITHUB_REF_NAME#ledgerexporter-v}
```
Comment (urvisavla) [inline .github/workflows/ledgerexporter-release.yml:17]:
```suggestion
      VERSION: ${GITHUB_REF_NAME#ledgerexporter-v} 
```

File: exp/services/ledgerexporter/Makefile [hunk 6]
```diff
@@ -14,32 +14,6 @@ $(if $(STELLAR_CORE_VERSION), --build-arg STELLAR_CORE_VERSION=$(STELLAR_CORE_VE
 	-t $(DOCKER_IMAGE):$(VERSION) \
 	-t $(DOCKER_IMAGE):latest .
 
-docker-clean:
-	$(SUDO) docker stop fake-gcs-server || true
-	$(SUDO) docker rm fake-gcs-server || true
-	$(SUDO) rm -rf ${PWD}/storage || true
-	$(SUDO) docker network rm test-network || true
-    
-docker-test: docker-clean
```
Comment (urvisavla) [inline exp/services/ledgerexporter/Makefile:23]: Can we keep this in? Even though it's no longer directly used in gh actions, it serves as a good example of how to locally run the ledger exporter docker with the fake-gcs-serve without digging through the code.

File: exp/services/ledgerexporter/README.md (hunk 2 above)
Comment (sreuland) [reply exp/services/ledgerexporter/README.md]: ok, let me check, the [code does an ImagePull](https://github.com/stellar/go/pull/5370/files#diff-1c787ef8e522ba695cb843ceef6c93e842a3cfc319085d6b78349a1c4ec201abR188), but that seems to not be resulting in the image getting into local cache

Comment (sreuland) [reply exp/services/ledgerexporter/README.md]: ok, think I fixed this, https://github.com/stellar/go/pull/5370/commits/0d3a999fb1cb0e8bb35498bf06ea4bf3bbfe5615

was able to replicate and see it work after

Comment (urvisavla) [reply exp/services/ledgerexporter/README.md]: hm I'm seeing the same error as before.

File: exp/services/ledgerexporter/internal/integration_test.go [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+func (s *LedgerExporterTestSuite) TestAppend() {
+	require := s.Require()
+
+	// first populate ledgers 4-5
+	rootCmd.SetArgs([]string{"scan-and-fill", "--start", "4", "--end", "5", "--config-file", s.tempConfigFile})
+	err := rootCmd.ExecuteContext(s.ctx)
+	require.NoError(err)
+
+	// now run an append of overalapping range, it will resume past existing ledgers 4,5
+	rootCmd.SetArgs([]string{"append", "--start", "4", "--end", "7", "--config-file", s.tempConfigFile})
```
Comment (urvisavla) [inline exp/services/ledgerexporter/internal/integration_test.go:87]: Do we need to also test unbounded mode where the end ledger is not provided?

File: exp/services/ledgerexporter/internal/integration_test.go [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+			PortBindings: nat.PortMap{
+				nat.Port("1570/tcp"):  {nat.PortBinding{HostIP: "127.0.0.1", HostPort: "1570"}},
+				nat.Port("11625/tcp"): {nat.PortBinding{HostIP: "127.0.0.1", HostPort: "11625"}},
+			},
+			AutoRemove: true,
+		},
+		nil, nil, "")
+
+	if err != nil {
+		t.Fatalf("could not create quickstart docker container, %v, error %v", quickstartImage, err)
```
Comment (urvisavla) [inline exp/services/ledgerexporter/internal/integration_test.go:216]: Tests are failing with this error when running locally
    "Integration_test.go:216: could not create quickstart docker container, stellar/quickstart:testing, error Error response from daemon: No such image: stellar/quickstart:testing"

File: exp/services/ledgerexporter/Makefile (hunk 6 above)
Comment (sreuland) [reply exp/services/ledgerexporter/Makefile:23]: yes, renamed it slightly to `docker-test-fake-gcs` - https://github.com/stellar/go/pull/5370/commits/c75d9939913bd5c3851cfc235ec950ae12ca8041

File: exp/services/ledgerexporter/internal/integration_test.go (hunk 7 above)
Comment (sreuland) [reply exp/services/ledgerexporter/internal/integration_test.go]: yes, good point, added it - https://github.com/stellar/go/pull/5370/files#diff-1c787ef8e522ba695cb843ceef6c93e842a3cfc319085d6b78349a1c4ec201abR115

File: exp/services/ledgerexporter/internal/integration_test.go (hunk 8 above)
Comment (sreuland) [reply exp/services/ledgerexporter/internal/integration_test.go:296]: thanks for testing this locally, yes, I needed to fix another aspect of pulling using the sdk, s/b working now - https://github.com/stellar/go/pull/5370/commits/9374cf3fa406b92e8b63e600569d63a8d58211f4

**Review by urvisavla — APPROVED** [review]

**Pull request merged**